VRF and Gameplay contract update

AceVikings opened the pull request:
This update contains contract for Random number generation and token contract update for gameplay functionalities.

AceVikings committed:
Create RandomGenerator.sol

Contract to generate random number using Chainlink VRF

AceVikings committed:
Update CryptoMafiaCoin.sol

Added gameplay functionalities

Niraj-Kamdar commented:
Do discussed changes: 
createRoom - mapping random to room id , call request random for next room
getRandom - take room id and send mapped random back 
reward - playerId - token address, get owner from it and give them allowance
redeem reward - set allowance 0 and transfer it to user

AceVikings committed:
Added reward system and UX changes

Added reward function and function to change address for interface. Getting playerID from their ERC721 now and some efficiency changes

AceVikings committed:
Update CryptoMafiaCoin.sol

Added check for if random number already exist and delete it when game is over

Niraj-Kamdar commented on contracts/Accounts.sol:
```diff
@@ -2,15 +2,26 @@
+        TI.signUpReward(msg.sender, 100);
+    }
+    
+    function AddTIAddress(address tokenAddress) public {
+        TI = TokenInterface(tokenAddress);
```
This shouldn't be public.

AceVikings replied:
It has to be public but I'll change it to onlyOwner. We need to set the contract address after all the contracts have been deployed therefore this.

Niraj-Kamdar commented on contracts/CryptoMafiaCoin.sol:
```diff
@@ -4,12 +4,45 @@ pragma solidity ^0.8.0;
+    
+    AccountInterface AI;
+    
+    mapping(uint256=>uint256) roomStake;
```
What's the need for this?

AceVikings replied:
This tells us the total amount staked in the room for a given roomId

Niraj-Kamdar commented on contracts/CryptoMafiaCoin.sol:
```diff
@@ -4,12 +4,45 @@ pragma solidity ^0.8.0;
+    AccountInterface AI;
+    
+    mapping(uint256=>uint256) roomStake;
+    mapping(uint256=>mapping(uint256=>address)) playerID;
```
Still double mapping?

Niraj-Kamdar commented on contracts/CryptoMafiaCoin.sol:
```diff
@@ -4,12 +4,45 @@ pragma solidity ^0.8.0;
+    mapping(uint256=>uint256) roomStake;
+    mapping(uint256=>mapping(uint256=>address)) playerID;
+    mapping(uint256=>uint256) roomRandomNumber;
+    mapping(uint256=>bool) randomNumIsTaken;
```
No need. if first mapping is returning 0 then random number haven't fullfilled

AceVikings replied:
If we check key value mapping for just if a value exists for the key that doesn't guarantee that the value would be unique, basically I am check if a random number is already in use or not

Niraj-Kamdar replied:
We don't need unique random no. we just need a random no. for roomId. It's okay if sometime 2 random no. are equal

Niraj-Kamdar commented on contracts/CryptoMafiaCoin.sol:
```diff
@@ -4,12 +4,45 @@ pragma solidity ^0.8.0;
+    }
+    
+    
+    function setAccountAddress(address accountAddress) public {
+        AI = AccountInterface(accountAddress);
+    }
+    
+    function setRandomAddress(address accountAddress) public {
+        RNG = RandomInterface(accountAddress);
     }
 
```
Imagine a random person changing the addess and causing contract to misbehave. In worst case hack the contract and drain the fund

AceVikings replied:
I'll change it to onlyOwner

Niraj-Kamdar commented on contracts/RandomGenerator.sol:
```diff
@@ -0,0 +1,46 @@
+
+import "https://github.com/smartcontractkit/chainlink/blob/master/evm-contracts/src/v0.6/VRFConsumerBase.sol";
+
+contract RandomGenerator is VRFConsumerBase  {
```
Maintain roomId => randomNo mapping and requestId => roomid mapping here.

AceVikings replied:
Do we need requestId mapping ?

AceVikings replied:
roomId=>randomNo mapping already exists in GameCoin contract where its being used, any specific reason to move that here ?

Niraj-Kamdar replied:
Yes, if you don't there will be dirty update problem.

Niraj-Kamdar reviewed (changes requested):
Fix the suggested issues and maintaining roomId=> randomNo mapping is must for correct behaviour.

AceVikings committed:
Change contains security updates

Modifier for all address setting functions has been changed to only owner and double mapping removed.

Niraj-Kamdar commented on contracts/CryptoMafiaCoin.sol:
```diff
@@ -24,4 +56,58 @@ contract CryptoMafiaCoin is ERC20, Ownable {
     }
+    
+    
+    function RequestRandomNumberCMC(uint256 seed) public returns(bytes32){
```
use a nonce as a seed.

Niraj-Kamdar commented on contracts/CryptoMafiaCoin.sol:
```diff
@@ -24,4 +56,58 @@ contract CryptoMafiaCoin is ERC20, Ownable {
+    
+    function joinRoom(uint256 roomId) public payable
+    {   
+        transfer(contractAddress,10);
+        roomStake[roomId] = roomStake[roomId] + 10;
```
why maintain roomStake? it will be constant 8*fee

AceVikings committed:
Utility changes

1. MafiaCoin contract doesn't pre-seed anymore
2. Room stake has been hard coded to 80 for now and that amount is directly distributed to winners
3. RequestRandomNumber now auto seeds the function with block hash of previous block
4. GetRandomNumber now requires roomId parameter to return random number value
5. Concurrent calls still get successfully stored in mapping according to request id and room number
6. Create Room function has been removed
7. Game over function has been removed

Niraj-Kamdar commented on contracts/CryptoMafiaCoin.sol:
```diff
@@ -24,4 +53,43 @@ contract CryptoMafiaCoin is ERC20, Ownable {
+    }
+    
+    
+    function signUpReward(address userAddress,uint256 amount) public {
+        _mint(userAddress,amount);
+    }
```
use mint function directly

Niraj-Kamdar commented on contracts/CryptoMafiaCoin.sol:
```diff
@@ -24,4 +53,43 @@ contract CryptoMafiaCoin is ERC20, Ownable {
     function mint(address receiver, uint amount) onlyAccountsHolder public {
         _mint(receiver, amount);
     }
```
this one directly in ERC721 with interface

Niraj-Kamdar commented on contracts/RandomGenerator.sol:
```diff
@@ -0,0 +1,74 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.6.0;
+
+import "https://github.com/smartcontractkit/chainlink/blob/master/evm-contracts/src/v0.6/VRFConsumerBase.sol";
```
use this github URL isn't reliable, import "@chainlink/contracts/src/v0.6/VRFConsumerBase.sol";

AceVikings committed:
Completed requested changes
